Error installing mihari: How to Fix

Hey there! Wanted to try out your tool but having a few issues.

Was having issues with both the gem install and Docker container so I pulled down the most recent Ruby Docker container for a clean environment with:

`docker pull ruby:slim-buster`
`docker run -it ruby:slim-buster /bin/bash`

Once I have a shell inside of the container, I check the version of Ruby:
```
ruby --version
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux]
```

After running the install command below using gem, I have the proceeding output:

```
root@c49d2b8dd776:/# gem install mihari

Fetching i18n-1.8.2.gem
Fetching tzinfo-1.2.7.gem
Fetching zeitwerk-2.3.0.gem
Fetching concurrent-ruby-1.1.6.gem
Fetching activemodel-6.0.3.gem
Fetching activesupport-6.0.3.gem
Fetching rack-2.2.2.gem
Fetching thread_safe-0.3.6.gem
Fetching rack-test-1.1.0.gem
Fetching mini_portile2-2.4.0.gem
Fetching nokogiri-1.10.9.gem
Fetching crass-1.0.6.gem
Fetching loofah-2.5.0.gem
Fetching rails-html-sanitizer-1.3.0.gem
Fetching rails-dom-testing-2.0.3.gem
Fetching builder-3.2.4.gem
Fetching erubi-1.9.0.gem
Fetching public_suffix-4.0.5.gem
Fetching actionview-6.0.3.gem
Fetching actionpack-6.0.3.gem
Fetching jsonapi-renderer-0.2.2.gem
Fetching case_transform-0.2.gem
Fetching active_model_serializers-0.10.10.gem
Fetching activerecord-6.0.3.gem
Fetching addressable-2.7.0.gem
Fetching binaryedge-0.1.0.gem
Fetching censu-0.2.0.gem
Fetching pg-1.2.3.gem
Fetching crtsh-rb-0.3.0.gem
Fetching dnpedia-0.1.0.gem
Fetching dnstwister-0.1.0.gem
Fetching unf_ext-0.0.7.7.gem
Fetching unf-0.1.4.gem
Fetching netaddr-2.0.4.gem
Fetching email_address-0.1.17.gem
Fetching hachi-0.3.1.gem
Fetching mem-0.1.5.gem
Fetching misp-0.1.3.gem
Fetching murmurhash3-0.1.6.gem
Fetching simpleidn-0.1.1.gem
Fetching net-ping-2.0.7.gem
Fetching onyphe-2.0.0.gem
Fetching parallel-1.19.1.gem
Fetching passive_circl-0.1.0.gem
Fetching passivetotalx-0.1.1.gem
Fetching thor-1.0.1.gem
Fetching pulsedive-0.1.5.gem
Fetching securitytrails-1.0.0.gem
Fetching shodanx-0.2.1.gem
Fetching mihari-1.1.0.gem
Fetching slack-notifier-2.3.2.gem
Fetching sqlite3-1.4.2.gem
Fetching urlscan-0.5.0.gem
Fetching virustotalx-1.1.0.gem
Fetching zoomeye-rb-0.1.1.gem
Successfully installed concurrent-ruby-1.1.6

HEADS UP! i18n 1.1 changed fallbacks to exclude default locale.
But that may break your application.

If you are upgrading your Rails application from an older version of Rails:

Please check your Rails app for 'config.i18n.fallbacks = true'.
If you're using I18n (>= 1.1.0) and Rails (< 5.2.2), this should be
'config.i18n.fallbacks = [I18n.default_locale]'.
If not, fallbacks will be broken in your app by I18n 1.1.x.

If you are starting a NEW Rails application, you can ignore this notice.

For more info see:
https://github.com/svenfuchs/i18n/releases/tag/v1.1.0

Successfully installed i18n-1.8.2
Successfully installed thread_safe-0.3.6
Successfully installed tzinfo-1.2.7
Successfully installed zeitwerk-2.3.0
Successfully installed activesupport-6.0.3
Successfully installed activemodel-6.0.3
Successfully installed rack-2.2.2
Successfully installed rack-test-1.1.0
Successfully installed mini_portile2-2.4.0
Building native extensions. This could take a while...
ERROR:  Error installing mihari:
	ERROR: Failed to build gem native extension.

    current directory: /usr/local/bundle/gems/nokogiri-1.10.9/ext/nokogiri
/usr/local/bin/ruby -I /usr/local/lib/ruby/2.7.0 -r ./siteconf20200515-9-ob9jmn.rb extconf.rb
checking if the C compiler accepts ... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/usr/local/bin/$(RUBY_BASE_NAME)
	--help
	--clean
/usr/local/lib/ruby/2.7.0/mkmf.rb:471:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /usr/local/lib/ruby/2.7.0/mkmf.rb:597:in `block in try_compile'
	from /usr/local/lib/ruby/2.7.0/mkmf.rb:544:in `with_werror'
	from /usr/local/lib/ruby/2.7.0/mkmf.rb:597:in `try_compile'
	from extconf.rb:138:in `nokogiri_try_compile'
	from extconf.rb:162:in `block in add_cflags'
	from /usr/local/lib/ruby/2.7.0/mkmf.rb:655:in `with_cflags'
	from extconf.rb:161:in `add_cflags'
	from extconf.rb:416:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /usr/local/bundle/extensions/x86_64-linux/2.7.0/nokogiri-1.10.9/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /usr/local/bundle/gems/nokogiri-1.10.9 for inspection.
Results logged to /usr/local/bundle/extensions/x86_64-linux/2.7.0/nokogiri-1.10.9/gem_make.out
```

I tried installing `build-essential` and ran again.
Reading the `mkmf.log` file on the second attempt at install results in the following output:

```
root@c49d2b8dd776:/# cat /usr/local/bundle/extensions/x86_64-linux/2.7.0/pg-1.2.3/mkmf.log
find_executable: checking for pg_config... -------------------- no

--------------------

find_header: checking for libpq-fe.h... -------------------- no

"gcc -o conftest -I/usr/local/include/ruby-2.7.0/x86_64-linux -I/usr/local/include/ruby-2.7.0/ruby/backward -I/usr/local/include/ruby-2.7.0 -I.    -O3 -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable  -fPIC conftest.c  -L. -L/usr/local/lib -Wl,-rpath,/usr/local/lib -L. -fstack-protector-strong -rdynamic -Wl,-export-dynamic     -Wl,-rpath,/usr/local/lib -L/usr/local/lib -lruby  -lm   -lc"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

"gcc -E -I/usr/local/include/ruby-2.7.0/x86_64-linux -I/usr/local/include/ruby-2.7.0/ruby/backward -I/usr/local/include/ruby-2.7.0 -I.    -O3 -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable  -fPIC  conftest.c -o conftest.i"
conftest.c:3:10: fatal error: libpq-fe.h: No such file or directory
 #include <libpq-fe.h>
          ^~~~~~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: #include <libpq-fe.h>
/* end */

--------------------
```

Installed libpq-dev because the log shown above says we are missing, `libpq-fe.h`

`sudo apt-get install libpq-dev`

Attempted to install again and it failed. Reading the `mkmf.log` produced the following:

```
root@c49d2b8dd776:/# cat /usr/local/bundle/extensions/x86_64-linux/2.7.0/sqlite3-1.4.2/mkmf.log
package configuration for sqlite3 is not found
find_header: checking for sqlite3.h... -------------------- no

"gcc -o conftest -I/usr/local/include/ruby-2.7.0/x86_64-linux -I/usr/local/include/ruby-2.7.0/ruby/backward -I/usr/local/include/ruby-2.7.0 -I.    -O3 -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable  -fPIC conftest.c  -L. -L/usr/local/lib -Wl,-rpath,/usr/local/lib -L. -fstack-protector-strong -rdynamic -Wl,-export-dynamic     -Wl,-rpath,/usr/local/lib -L/usr/local/lib -lruby  -lm   -lc"
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

"gcc -E -I/usr/local/include/ruby-2.7.0/x86_64-linux -I/usr/local/include/ruby-2.7.0/ruby/backward -I/usr/local/include/ruby-2.7.0 -I.    -O3 -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-cond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrough=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable  -fPIC  conftest.c -o conftest.i"
conftest.c:3:10: fatal error: sqlite3.h: No such file or directory
 #include <sqlite3.h>
          ^~~~~~~~~~~
compilation terminated.
checked program was:
/* begin */
1: #include "ruby.h"
2:
3: #include <sqlite3.h>
/* end */

--------------------
```

Installed missing library again using apt:

`apt install sqlite3 libsqlite3-dev`

Ran the gem install again and now have a successful install.

```
root@c49d2b8dd776:/# gem install mihari
Successfully installed urlscan-0.5.0
Successfully installed virustotalx-1.1.0
Successfully installed zoomeye-rb-0.1.1
Successfully installed mihari-1.1.0
4 gems installed
```

While I was able to get this working, hoping this issue and the corresponding fix can be added to help some new users out. Please let me know if you have any questions or concerns. Have a good one, and keep up the good work!

--- a/docker/Dockerfile
+++ b/docker/Dockerfile
@@ -6,7 +6,7 @@ RUN apk --no-cache add git build-base ruby-dev sqlite-dev postgresql-dev \
   && gem build mihari.gemspec -o mihari.gem \
   && gem install mihari.gem \
   && rm -rf /tmp/mihari \
-  && apk del --purge git build-base ruby-dev sqlite-dev postgresql-dev
+  && apk del --purge git build-base ruby-dev
 
 ENTRYPOINT ["mihari"]
 

